proposals/nnnn-rpc-design-refactoring.md

```diff
@@ -0,0 +1,153 @@
+
+Now Application manager is a service locator and should to be passed to each component
+even if component does not need application manager, but need some other classes that owned by application manager.
+During construction components should explicitly receive all dependencies. 
+It will prevent from unwanted access and improve encapsulation. 
+
+
+#### Reduce responsibility of ApplicationManager class 
+Currently application manager class has too many responsibilities:
+ - Service location
+ - Mobile/HMI RPC processing 
+ - Mobile/HMI RPC sending 
+ - Registration and managing mobile applications
+ - Handle different states of SDL (low voltage/ video streaming / audio_path_throw ...) 
+ 
+ The purpose of this refactoring is to extract handling/sending RPC's to separate module.  
+ Many SDL commands require application manager only for sending RPC's and do not need to know anything related to other application manager responsibility. 
+ Also, extracting RPC processing to separate module will reduce regression risk of changing any functionality that now in ApplicationManager.
+ 
+#### Provide plugin architecture for easy adding additional functionality and RPC groups   
+
+SDL RPC's can be splitted to some groups or RPC's related to certain functionality (for example RC or video streaming).  
+Usually these RPC's work with some functionality that is not required for other RPC's, so should not be exposed. 
+
+Adding new RPC or group of RPC's is now very complicated process, but it is also not very obvious. 
+If OEM vendors need to introduce in their custom SDL forks new functionality, they need to change SDL code, 
+add new classes and modify existing factories. 
+
+The best solution would be plugin approach of adding new RPC's.
+So if some vendors need to implement code of new RPC's it would be enough to implement certain interfaces and provide them to SDL as shared library.
+
+
+## Detailed design
+
+### Use dependency injection for commands 
+
+Commands should be accepted as list of services that they require. And Mobile factory should be responsible for providing this list of dependencies.  
+
+#### Current Approach 
+```cpp
+
+SomeCommand::SomeCommand(app_mngr);
+
+void SomeCommand::Run() {
+  application_manager->protocol_handler().do_staff();
+  application_manager->state_controller().do_staff();
+  application_manager->PolicyHandler().do_staff();
+  application_manager->resume_controller().do_staff();
+  
+}
+```
+
+#### New Approach 
+```cpp
+
+SomeCommand::SomeCommand(protocol_handler, state_controller, policy_handler, resume_controller);
+
+void SomeCommand::Run() {
+  protocol_handler_.do_staff();
+  state_controller_.do_staff();
+  protocol_handler_.do_staff();
+  protocol_handler_.do_staff();
+}
+```
+This will simplify code and provide possibility to add a new component to certain commands without affecting other commands.
+And factory will create command with all required parameters, so existing code will not be complicated.  
```

Comment:
The factory will create the command with all required parameters, so existing code will not be complicated.

Reply:
@mrapitis, done in 2c9f7a3f7bb7e6c8b26bb1e4aec779cf65733666